core/error.rs

```diff
@@ -1,143 +1,260 @@
-        .map(|_| "TypeError")
-    })
-}
+  fn get_message(&self) -> Cow<'static, str> {
+    match self {
+      CoreError::Js(js_error) => {
+        if let Some(name) = &js_error.message {
+          Cow::Owned(name.clone())
+        } else {
+          Cow::Borrowed("")
+        }
+      }
+      CoreError::Io(err) => err.get_message(),
+      CoreError::ExtensionTranspiler(err) => err.get_message(),
+      CoreError::ModuleLoader(err) => err.get_message(),
+      CoreError::CouldNotExecute { error, .. } => error.get_message(),
+      CoreError::JsNative(err) => err.get_message(),
+      CoreError::Url(err) => err.get_message(),
+      CoreError::Module(err) => err.get_message(),
+      CoreError::DataError(err) => err.get_message(),
+      CoreError::TLA
+      | CoreError::Parse(_)
+      | CoreError::Execute(_)
+      | CoreError::UnusedModules(_)
+      | CoreError::NonEvaluatedModules(_)
+      | CoreError::MissingFromModuleMap(_)
+      | CoreError::FutureCanceled(_)
+      | CoreError::ExecutionTerminated
+      | CoreError::PendingPromiseResolution
+      | CoreError::EvaluateDynamicImportedModule
+      | CoreError::Other(_) => self.to_string().into(),
+    }
+  }
 
-/// A wrapper around `anyhow::Error` that implements `std::error::Error`
-#[repr(transparent)]
-pub struct StdAnyError(pub Error);
-impl std::fmt::Debug for StdAnyError {
-  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-    write!(f, "{:?}", self.0)
+  fn get_additional_properties(
+    &self,
+  ) -> Vec<(Cow<'static, str>, Cow<'static, str>)> {
+    vec![] // TODO
   }
-}
 
-impl std::fmt::Display for StdAnyError {
-  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-    write!(f, "{}", self.0)
+  fn as_any(&self) -> &dyn Any {
+    self
   }
 }
 
-impl std::error::Error for StdAnyError {
-  fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-    self.0.source()
-  }
+pub fn throw_js_error_class<E: JsErrorClass>(
+  scope: &mut v8::HandleScope,
+  error: &E,
+) {
+  let exception = js_class_and_message_to_exception(
+    scope,
+    &error.get_class(),
+    &error.get_message(),
+  );
+  scope.throw_exception(exception);
 }
 
-impl From<Error> for StdAnyError {
-  fn from(err: Error) -> Self {
-    Self(err)
+fn js_class_and_message_to_exception<'s>(
+  scope: &mut v8::HandleScope<'s>,
+  class: &str,
+  message: &str,
+) -> v8::Local<'s, v8::Value> {
+  let message = v8::String::new(scope, message).unwrap();
+  match class {
+    TYPE_ERROR => v8::Exception::type_error(scope, message),
+    RANGE_ERROR => v8::Exception::range_error(scope, message),
+    REFERENCE_ERROR => v8::Exception::reference_error(scope, message),
+    SYNTAX_ERROR => v8::Exception::syntax_error(scope, message),
+    _ => v8::Exception::error(scope, message),
   }
 }
 
 pub fn to_v8_error<'a>(
   scope: &mut v8::HandleScope<'a>,
-  get_class: GetErrorClassFn,
-  error: &Error,
+  error: &impl JsErrorClass,
```

Comment:
If it doesn't hurt perf, maybe use `dyn` instead of `impl` here so this function isn't duplicated for each error type? May not be a big deal though

Reply:
good point; done. found another spot where i could do this as well so done there too.